internal/campaigns/service.go

```diff
@@ -415,6 +415,70 @@ func (svc *Service) ExecuteCampaignSpec(ctx context.Context, opts ExecutorOpts,
 	return specs, x.LogFiles(), errs.ErrorOrNil()
 }
 
+func (svc *Service) ValidateChangesetSpecs(repos []*graphql.Repository, specs []*ChangesetSpec) error {
+	repoByID := make(map[string]*graphql.Repository, len(repos))
+	for _, repo := range repos {
+		repoByID[repo.ID] = repo
+	}
+
+	byRepoAndBranch := make(map[string]map[string][]*ChangesetSpec)
+	for _, spec := range specs {
+		_, ok := byRepoAndBranch[spec.HeadRepository]
+		if !ok {
+			byRepoAndBranch[spec.HeadRepository] = make(map[string][]*ChangesetSpec)
+		}
+
+		byRepoAndBranch[spec.HeadRepository][spec.HeadRef] = append(byRepoAndBranch[spec.HeadRepository][spec.HeadRef], spec)
+	}
+
+	duplicates := make(map[*graphql.Repository]map[string]int)
+	for repoID, specsByBranch := range byRepoAndBranch {
+		for branch, specs := range specsByBranch {
+			if len(specs) < 2 {
+				continue
+			}
+
+			r := repoByID[repoID]
+			if _, ok := duplicates[r]; !ok {
+				duplicates[r] = make(map[string]int)
+			}
+
+			duplicates[r][branch] = len(specs)
+		}
+	}
+
+	if len(duplicates) > 0 {
+		return &duplicateBranchErr{fooduplicates: duplicates}
```

Comment:
`fooduplicates`? 😆

Reply:
I mean, that _is_ a unique name. :smiley:

Reply:
Argh. I gotta clean up my stuff better.